### PR TITLE
Implement type pooling for `T::Types::TypedArray`

### PR DESCRIPTION
**Pull request by casperisfine**
From heap-profiler:

```
retained objects by class
-----------------------------------
     11416  T::Types::TypedArray
```

cc @paracycle @jez

### CODE REVIEW

File: gems/sorbet-runtime/lib/sorbet-runtime.rb [hunk 1]
```diff
@@ -83,6 +82,9 @@ module T::Private::Types; end
 require_relative 'types/utils'
 require_relative 'types/boolean'
 
+# Depdends on types/utils
```
Comment (paracycle) [inline gems/sorbet-runtime/lib/sorbet-runtime.rb:85]:
```suggestion
# Depedends on types/utils
```

**Comment by froydnj** [comment]
> From heap-profiler:
> 
> ```
> retained objects by class
> -----------------------------------
>      11416  T::Types::TypedArray
> ```

Can you show how this PR changes the above number?  It's not clear to me whether this is before or after, or what codebase this number is coming from.

**Comment by casperisfine** [comment]
> It's not clear to me whether this is before or after, or what codebase this number is coming from.

So it's before, and from Shopify's monolith codebase. I'm yet to try this patch, but the assumption is that a majority of these are common types such as `T::Array[String]` etc.

**Comment by casperisfine** [comment]
> I'm yet to try this patch

Hum, it's unclear how to point a Gemfile to a sorbet branch, since the gemspecs set `s.version = '0.0.0'` bundler refuses to use it :/

**Comment by casperisfine** [comment]
Nevermind, I figured a way to hack this. I just pushed another branch that also modify the gemspec.

**Comment by casperisfine** [comment]
Ok, so it was effective enough that it no longer makes it in the top 50, so I had to grep the heap dump manually:

```
$ rg -F '"T::Types::TypedArray"' ~/Downloads/shopify-production-boot-heap-2023-04-06.dump 
6079158:{"address":"0x7f92a053fa68", "type":"STRING", "shape_id":12, "slot_size":80, "class":"0x7f92a627eda0", "frozen":true, "embedded":true, "fstring":true, "bytesize":20, "value":"T::Types::TypedArray", "encoding":"US-ASCII", "coderange":"7bit", "file":"/tmp/bundle/ruby/3.2.0/bundler/gems/sorbet-b50de5e3839a/gems/sorbet-runtime/lib/types/types/typed_array.rb", "line":68, "generation":51, "memsize":80, "flags":{"wb_protected":true, "old":true, "uncollectible":true, "marked":true}}
7374949:{"address":"0x7f92a12efa00", "type":"CLASS", "shape_id":13, "slot_size":160, "class":"0x7f92a12ef8c0", "variation_count":0, "superclass":"0x7f92a0cb7db8", "name":"T::Types::TypedArray", "references":["0x7f92a0cb7db8", "0x7f92a1eba1f0", "0x7f92a1eba588", "0x7f92a1eba308", "0x7f92a1eba1a0", "0x7f92a1eba330", "0x7f92a1eba588", "0x7f9132105fe8", "0x7f913226d750", "0x7f9121d2e470", "0x7f92a801e928", "0x7f91322af498", "0x7f92a0437558", "0x7f913226d9a8", "0x7f92a04375d0", "0x7f913226d548", "0x7f91f845f830", "0x7f92a043ba90", "0x7f9132106010", "0x7f9134a28718", "0x7f92a043ba68", "0x7f913226d778", "0x7f92a043bb80", "0x7f917c4db218", "0x7f91230e3678", "0x7f92a8033f80", "0x7f91321062b8", "0x7f92a043ba18", "0x7f91bc6e2710", "0x7f91344caff0", "0x7f9123115bc8", "0x7f92a053fa68", "0x7f92a12ef820", "0x7f92a03a3a88", "0x7f92a12ef640", "0x7f92a03a3a88"], "file":"/tmp/bundle/ruby/3.2.0/bundler/gems/sorbet-b50de5e3839a/gems/sorbet-runtime/lib/types/types/typed_array.rb", "line":68, "generation":51, "memsize":1520, "flags":{"wb_protected":true, "old":true, "uncollectible":true, "marked":true}}
$ rg -F '"class":"0x7f92a12efa00"' ~/Downloads/shopify-production-boot-heap-2023-04-06.dump | wc -l
    2804
```

So it's down from `11416` to `2804` instances in our app.

**Review by jez — APPROVED** [review]

**Comment by jez** [comment]
http://go/builds/bui_NffNxN6ugwxiq2

**Pull request merged**